### PR TITLE
[tech-onboarding assessments] Implement key server which can generate random api keys

### PR DESCRIPTION
**Pull request by abhinvv1**
### Why?

- This PR is to implement a Redis-based API key server which can generate random api keys, assign them for usage and release them after sometime.
* Project doc: https://docs.google.com/document/d/11ORDg227vZiHJBtOaxSGxTB8tG9D8jDdIYOsoGMv7K8/edit?usp=sharing 

### Key Features:
1. Generation of unique API keys with expiration
2. Key allocation and blocking mechanism
3. Unblocking of keys after a set timer
4. Key deletion and keep-alive functionality
5. Background process for managing key expiration and unblocking

### Implementation Details:
1. Uses Redis for data storage with sets and sorted sets for O(1) or O(log N) time complexity
2. Sinatra-based REST API for key management operations
3. RSpec tests for all key functionalities

- NF requirements:
  - RSpecs (Unit tests) expected for the code.
  - Should build this without full fledged Rails framework.

### Testing:
- RSpec test suite covering all key operations 
- run `rspec` to run the tests

### Usage
1. create .env file with REDIS_URL=redis://localhost:6379
2. start redis on port 6379
3. to run the server: ruby api_key_server.rb
4. to run the tests: rspec

Generate new keys:
```
curl --location 'http://localhost:4567/keys' \
--header 'Content-Type: application/json' \
--data ''
```
Get the keys:

`curl --location 'http://localhost:4567/keys/available'`

Delete the keys:
```
curl --location --request DELETE 'http://localhost:4567/keys?key=<key-value>' \
--data ''
```
Unblock keys:
```
curl --location --request PATCH 'http://localhost:4567/keys/<key-value>/unblock' \
--header 'Content-Type: application/json' \
--data ''
```

### Sample output
- Generate keys: 
Screenshot 2024-07-26 at 5 38 50 PM

- Get keys:
Screenshot 2024-07-26 at 5 39 01 PM

- Unblock keys:
Screenshot 2024-07-26 at 5 39 22 PM

- Delete keys:
Screenshot 2024-07-26 at 5 39 37 PM

### CODE REVIEW

File: routes/routes.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      status 404
+      { error: 'Key not found' }.to_json
+    end
+  end
+
+  delete '/keys' do
+    key = params[:key]
+    @key_manager.delete_key(key)
+    { message: 'Key deleted' }.to_json
+  end
```
Comment (gotunandan) [inline routes/routes.rb:46]: This returns 200 OK even if you provide a random key to delete, which does not exist, or has already been deleted.
It should return 404

Comment (gotunandan) [reply routes/routes.rb:46]: Add a test case for this too

**Review by gotunandan — CHANGES_REQUESTED** [review]
Couple of small fixes + one bug in the `/delete` endpoint

Please check

File: routes/routes.rb (hunk 1 above)
Comment (abhinvv1) [reply routes/routes.rb]: thanks for pointing this out, I've updated the `key_manager.rb` and `routes.rb`, also added tests for the same in `api_key_server_spec.rb`

**Review by gotunandan — APPROVED** [review]
LGTM

**Pull request merged**